File: examples/nodejs/verify_tls.js
/**
 * verify_tls.js — TLS Certificate Chain Verification (Node.js)
 * CSE PKI — University of Dhaka
 *
 * Demonstrates client-side TLS verification against a private Root CA
 * using Node.js built-in 'https' module.
 *
 * Run: node verify_tls.js
 */

"use strict";

const https = require("https");
const fs    = require("fs");
const path  = require("path");

// Path to our private Root CA certificate
// Clients MUST explicitly configure this — not in OS/browser trust stores
const ROOT_CA_CERT = path.resolve(__dirname, "../../certs/root/ca.cert.pem");

const BASE_HOSTNAME = "iot-cse.du.ac.bd";
const BASE_PORT     = 443;


/**
 * Make an HTTPS GET request, verifying against our private Root CA.
 */
function httpsGet(hostname, port, urlPath, extraOptions = {}) {
  return new Promise((resolve, reject) => {
    const options = {
      hostname,
      port,
      path: urlPath,
      method: "GET",
      ca: fs.readFileSync(ROOT_CA_CERT),   // Custom CA — replaces system trust store
      rejectUnauthorized: true,            // ALWAYS true in production
      ...extraOptions,
    };

    const req = https.request(options, (res) => {
      let data = "";
      res.on("data", (chunk) => { data += chunk; });
      res.on("end", () => resolve({ statusCode: res.statusCode, body: data }));
    });

    req.on("error", reject);
    req.setTimeout(10_000, () => {
      req.destroy(new Error("Request timed out"));
    });
    req.end();
  });
}


async function checkStatus() {
  console.log(`\n[1] GET https://${BASE_HOSTNAME}/status`);
  try {
    const { statusCode, body } = await httpsGet(BASE_HOSTNAME, BASE_PORT, "/status");
    console.log(`    Status : ${statusCode}`);
    console.log(`    Body   : ${body.slice(0, 200)}`);
  } catch (err) {
    console.log(`    [WARN] ${err.message} (server may not be running in demo env)`);
  }
}


async function checkMultipleEndpoints() {
  console.log(`\n[2] Checking multiple endpoints...`);
  const endpoints = ["/status", "/api/health", "/api/sensors"];

  for (const endpoint of endpoints) {
    try {
      const { statusCode } = await httpsGet(BASE_HOSTNAME, BASE_PORT, endpoint);
      console.log(`    ${endpoint.padEnd(20)} → ${statusCode}`);
    } catch (err) {
      console.log(`    ${endpoint.padEnd(20)} → [${err.code || err.message}]`);
    }
  }
}


async function demonstrateInsecure() {
  console.log(`\n[3] Demonstrating rejectUnauthorized: false (INSECURE — demo only)`);
  try {
    const result = await httpsGet(
      BASE_HOSTNAME, BASE_PORT, "/status",
      { rejectUnauthorized: false }
    );
    console.log(`    Status: ${result.statusCode}`);
    console.log(`    [WARN] rejectUnauthorized=false disables ALL certificate checks.`);
    console.log(`    [WARN] NEVER use this in production — it enables MITM attacks.`);
  } catch (err) {
    console.log(`    [not reachable — expected in demo environment]`);
  }
}


async function main() {
  console.log("=".repeat(60));
  console.log("  CSE PKI — Node.js TLS Verification Demo");
  console.log("=".repeat(60));

  if (!fs.existsSync(ROOT_CA_CERT)) {
    console.warn(`\n  [WARN] Root CA cert not found at: ${ROOT_CA_CERT}`);
    console.warn("  Run scripts/01_build_root_ca.sh first.\n");
  } else {
    console.log(`  Root CA: ${ROOT_CA_CERT}`);
  }

  await checkStatus();
  await checkMultipleEndpoints();
  await demonstrateInsecure();

  console.log("\nDone.\n");
}

main().catch(console.error);

File: examples/python/verify_tls.py
"""
verify_tls.py — TLS Certificate Chain Verification (Python)
CSE PKI — University of Dhaka

Demonstrates client-side TLS verification against a private Root CA
using the 'requests' library.

Install:  pip install requests
Run:      python verify_tls.py
"""

import sys
import requests
from requests.exceptions import SSLError, ConnectionError

# Path to our private Root CA certificate
# Clients MUST explicitly trust this CA (not in browser/OS stores)
ROOT_CA_CERT = "../../certs/root/ca.cert.pem"

BASE_URL = "https://iot-cse.du.ac.bd"


def check_status() -> None:
    """Simple GET with CA verification."""
    print(f"\n[1] GET {BASE_URL}/status")
    try:
        response = requests.get(
            f"{BASE_URL}/status",
            verify=ROOT_CA_CERT,   # <-- chain validated against our Root CA
            timeout=10,
        )
        response.raise_for_status()
        print(f"    Status : {response.status_code}")
        print(f"    Body   : {response.text[:200]}")
    except SSLError as e:
        print(f"    [FAIL] SSL/TLS error: {e}")
        sys.exit(1)
    except ConnectionError as e:
        print(f"    [WARN] Could not connect (server may not be running): {e}")


def check_with_session() -> None:
    """Reuse a session across multiple requests (efficient for IoT polling)."""
    print(f"\n[2] Session-based requests to {BASE_URL}")
    session = requests.Session()
    session.verify = ROOT_CA_CERT

    endpoints = ["/status", "/api/sensors", "/api/health"]
    for endpoint in endpoints:
        try:
            response = session.get(f"{BASE_URL}{endpoint}", timeout=5)
            print(f"    {endpoint:20s} → {response.status_code}")
        except ConnectionError:
            print(f"    {endpoint:20s} → [not reachable]")


def check_wrong_ca() -> None:
    """Show that verification fails if we use the wrong CA (security check)."""
    print(f"\n[3] Verifying that wrong CA triggers rejection...")
    try:
        response = requests.get(
            f"{BASE_URL}/status",
            verify=False,  # <-- INSECURE: disables verification
        )
        print("    [WARN] verify=False bypasses all certificate checks — NEVER do this in production!")
    except ConnectionError:
        print("    [not reachable — expected in demo environment]")


if __name__ == "__main__":
    print("=" * 60)
    print("  CSE PKI — Python TLS Verification Demo")
    print("=" * 60)
    print(f"  Root CA: {ROOT_CA_CERT}")

    check_status()
    check_with_session()
    check_wrong_ca()

    print("\nDone.\n")
